File: JavaFundamentals/ExamPractice/src/LegendaryFarming.java
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LegendaryFarming {
    public static void main(String[] args) {
        Scanner Console = new Scanner(System.in);
        boolean breakWhile = false;
        TreeMap<String, Integer> materials = new TreeMap<>();
        materials.put("fragments", 0);
        materials.put("motes", 0);
        materials.put("shards", 0);

        TreeMap<String, Integer> junks = new TreeMap<>();
        String input = Console.nextLine().toLowerCase();
        Pattern pat = Pattern.compile("(\\d+)\\s+([\\w-]+)");
        StringBuilder sb = new StringBuilder();

        while (true) {
            Matcher match = pat.matcher(input);
            while (match.find()) {
                int quantity = Integer.parseInt(match.group(1));
                String mats = match.group(2);

                if (mats.equals("fragments") || mats.equals("motes") || mats.equals("shards")) {
                    materials.put(mats, materials.get(mats) + quantity);
                    if (materials.get(mats) >= 250) {
                        materials.put(mats, materials.get(mats) - 250);
                        sb.append(mats);

                        breakWhile = true;
                        break;
                    }

                } else {
                    if (!junks.containsKey(mats)) {
                        junks.put(mats, quantity);

                    } else {
                        junks.put(mats, junks.get(mats) + quantity);
                    }
                }
            }
            if (breakWhile) {
                break;
            }
            input = Console.nextLine().toLowerCase();
        }

        String sbString = sb.toString();
        if (sbString.equals("motes")) {
            System.out.println("Dragonwrath obtained!");
        } else if (sbString.equals("fragments")) {
            System.out.println("Valanyr obtained!");
        } else {
            System.out.println("Shadowmourne obtained!");
        }

        materials.entrySet()
                .stream()
                .sorted((a, b) -> b.getValue().compareTo(a.getValue()))
                .forEach(e ->
                        System.out.println(e.getKey() + ": " + e.getValue()));
        junks.entrySet()
                .stream()
                .forEach(e ->
                        System.out.println(e.getKey() + ": " + e.getValue()));
    }
}

File: AdvancedCSharp/ExamPractice/17.LegendaryFarming/LegendaryFarming.cs
namespace _17.LegendaryFarming
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    internal class LegendaryFarming
    {
        private static void Main()
        {
            var junk = new SortedDictionary<string, int>();
            var keyMats = new Dictionary<string, int>();
            keyMats.Add("shards", 0);
            keyMats.Add("motes", 0);
            keyMats.Add("fragments", 0);

            var keyWinner = string.Empty;

            while (keyWinner == String.Empty)
            {
                var input = Console.ReadLine().Split(' ');

                for (int i = 0; i < input.Length; i+=2)
                {
                    var material = input[i + 1].ToLower();
                    var quantity = int.Parse(input[i]);

                    if (!keyMats.ContainsKey(material))
                    {
                        if (!junk.ContainsKey(material))
                        {
                            junk[material] = 0;
                        }
                        junk[material] += quantity;
                    }
                    else
                    {
                        keyMats[material] += quantity;
                        if (keyMats[material] >= 250)
                        {
                            keyMats[material] -= 250;
                            keyWinner += material;
                            break;
                        }
                    }
                }

            }

            if (keyWinner.Equals("motes"))
            {
                Console.WriteLine("Dragonwrath obtained!");
            }
            else if (keyWinner.Equals("fragments"))
            {
                Console.WriteLine("Valanyr obtained!");
            }
            else
            {
                Console.WriteLine("Shadowmourne obtained!");
            }

            var newKeyMats = keyMats.OrderByDescending(x => x.Value);
            foreach (var pair in newKeyMats)
            {
                Console.WriteLine("{0}: {1}", pair.Key , pair.Value);
            }
            foreach (var pair in junk)
            {
                Console.WriteLine("{0}: {1}", pair.Key, pair.Value);
            }
        }
    }
}